python/cudf_polars/cudf_polars/dsl/translate.py

```diff
@@ -19,10 +19,107 @@
 from polars.polars import _expr_nodes as pl_expr, _ir_nodes as pl_ir
 
 from cudf_polars.dsl import expr, ir
-from cudf_polars.typing import NodeTraverser
 from cudf_polars.utils import dtypes
 
-__all__ = ["translate_ir", "translate_named_expr"]
+if TYPE_CHECKING:
+    from cudf_polars.typing import NodeTraverser
+__all__ = ["Translator", "translate_named_expr"]
+
+
+class Translator:
+    """
+    Translates polars-internal IR nodes and expressions to our representation.
+
+    Parameters
+    ----------
+    visitor
+        Polars NodeTraverser object
+    debug_mode
+        Setting this mode to True allows Translator to collect
+        errors raised during translation
+    """
+
+    def __init__(self, visitor: NodeTraverser, *, debug_mode: bool = False):
+        self.visitor = visitor
+        self.debug_mode = debug_mode
+        self.errors: list[str] = []
+
+    def translate_ir(self, *, n: int | None = None) -> ir.IR:
+        """
+        Translate a polars-internal IR node to our representation.
+
+        Parameters
+        ----------
+        n
+            Optional node to start traversing from, if not provided uses
+            current polars-internal node.
+
+        Returns
+        -------
+        Translated IR object
+
+        Raises
+        ------
+        NotImplementedError
+            If we can't translate the nodes due to unsupported functionality.
+        """
+        ctx: AbstractContextManager[None] = (
+            set_node(self.visitor, n) if n is not None else noop_context
+        )
+        # IR is versioned with major.minor, minor is bumped for backwards
+        # compatible changes (e.g. adding new nodes), major is bumped for
+        # incompatible changes (e.g. renaming nodes).
+        # Polars 1.7 changes definition of the CSV reader options schema name.
+        if (version := self.visitor.version()) >= (3, 0):
+            raise NotImplementedError(
+                f"No support for polars IR {version=}"
+            )  # pragma: no cover; no such version for now.
+
+        with ctx:
+            polars_schema = self.visitor.get_schema()
+            node = self.visitor.view_current_node()
+            schema = {k: dtypes.from_polars(v) for k, v in polars_schema.items()}
+            try:
+                result = _translate_ir(node, self, schema)
+            except Exception as e:
+                self.errors.append(str(e))
+                return ir.ErrorNode(polars_schema, str(e))
+            if any(
+                isinstance(dtype, pl.Null)
+                for dtype in pl.datatypes.unpack_dtypes(*polars_schema.values())
+            ):
+                error = f"No GPU support for {result} with Null column dtype."
+                if self.debug_mode:
+                    self.errors.append(error)
+                    return ir.ErrorNode(polars_schema, error)
+                raise NotImplementedError(error)
```

Comment:
If we gate this behind `debug_mode` should we do the same on line 85-86?

However, my _preference_ is that unless it is prohibitively expensive, we should always collect all of the errors and then raise "late", so:

```suggestion
                self.errors.append(error)
                return ir.ErrorNode(schema, error)
```

Note, we should use the `schema` object, not the `polars_schema` object since otherwise we have the wrong type.